Code/GraphMol/Bond.cpp

```diff
@@ -291,6 +291,79 @@ void Bond::setStereoAtoms(unsigned int bgnIdx, unsigned int endIdx) {
   atoms.push_back(endIdx);
 };
 
+namespace {
+const Atom *findHighestCIPNeighbor(const Atom *atom, const Atom *skipAtom) {
+  unsigned bestCipRank = 0;
+  const Atom *bestCipRankedAtom = nullptr;
+  const auto &mol = atom->getOwningMol();
+
+  for (const auto &index :
+       boost::make_iterator_range(mol.getAtomNeighbors(atom))) {
+    const auto neighbor = mol[index];
+    if (neighbor == skipAtom) {
+      continue;
+    }
+    unsigned cip = 0;
+    if (!neighbor->getPropIfPresent(common_properties::_CIPRank, cip)) {
+      // If at least one of the atoms doesn't have a CIP rank, the highest rank
+      // does not make sense, so return a nullptr.
+      return nullptr;
+    }
+    if (cip > bestCipRank || bestCipRankedAtom == nullptr) {
+      bestCipRank = cip;
+      bestCipRankedAtom = neighbor;
+    } else if (cip == bestCipRank) {
+      // This also doesn't make sense if there is a tie (if that's possible).
+      // We still keep the best CIP rank in case something better comes around
+      // (also not sure if that's possible).
+      bestCipRankedAtom = nullptr;
+    }
+  }
+  return bestCipRankedAtom;
+}
+
+INT_VECT *findStereoAtoms(const Bond *bond) {
+  if (!bond->hasOwningMol()) {
+    return new INT_VECT;
+  }
+  switch (bond->getStereo()) {
+    case Bond::BondStereo::STEREOE:
+    case Bond::BondStereo::STEREOZ: {
+      const Atom *startStereoAtom =
+          findHighestCIPNeighbor(bond->getBeginAtom(), bond->getEndAtom());
+      const Atom *endStereoAtom =
+          findHighestCIPNeighbor(bond->getEndAtom(), bond->getBeginAtom());
+
+      if (startStereoAtom != nullptr && endStereoAtom != nullptr) {
+        int startStereoAtomIdx = static_cast<int>(startStereoAtom->getIdx());
+        int endStereoAtomIdx = static_cast<int>(endStereoAtom->getIdx());
+
+        return new INT_VECT({startStereoAtomIdx, endStereoAtomIdx});
+
+        break;
+      }
+    }
+    default:
+      return new INT_VECT;
+  }
+}
+
+}  // namespace
+
+const INT_VECT &Bond::getStereoAtoms() const {
+  if (!dp_stereoAtoms) {
+    const_cast<Bond *>(this)->dp_stereoAtoms = findStereoAtoms(this);
```

Comment:
We should just make dp_stereoAtoms mutable at least to remove the const_cast.

This also indicates that we need to put a lock here at some point, but that is probably a bridge too far.

Reply:
I disagree strongly. 
dp_stereoAtoms shouldn't be mutable any more than the bond order should be mutable. 
This is a special case where we're initializing a value that hasn't already been initialized (because it wasn't needed), it's not making an actual substantive change to the bond itself.

There's an argument to be made that the code in Chirality.cpp which assigns STEREOE/STEREOZ should set the stereo atoms so that we don't have to do this here, but that's a different point.

Reply:
after having watched the video about the new meaning of const that Brian recommended (https://www.google.com/amp/s/herbsutter.com/2013/01/01/video-you-dont-know-const-and-mutable/amp/) and thinking about it a bit more, I'm no longer thrilled with the fact that this getter is now no longer threadsafe.

Fixing this is going to require re-thinking the way we handle `dp_stereoAtoms`. I think this means creating the `dp_stereoAtoms` structure when we set STEREOE/STEREOZ and then having a PRECONDITION() in this method that fails if `dp_stereoAtoms` is not set.

Since that's a change in behavior beyond the way stereo in reactions is handled, I would be fine if we merged this PR as is and immediately created an issue for the threadsafety change. I can then take a swing at a solution for the problem. Alternatively, if you guys think it makes more sense to get it out of the way immediately, I can do a PR against @ricrogz's branch.

Reply:
I'm ok with both solutionsI don't have any preference.

Reply:
A third option would be to limit the scope of this change to bonds that are already have `stereoAtoms`, and leave resolution of `stereoAtoms` for E/Z bonds as a separate concern. I think that would preserve the thread-safety and `const` character of ``dp_stereoAtoms`, and address most of the problems that this commit is intending to address.

Then we could decide in some other commit whether stereoAtoms is a cache (and so calculated on the fly with some appropriate locking) or an attribute (and so calculated during sanitize or some other similar rectification step).

Reply:
@d-b-w, I think limiting the scope of this case to E/Z bonds which have stereoatoms set, or checking and setting them in every place they are required would be more complicated than addressing here, even if in a temporary manner.

I'll update with a temporary fix so we can merge this being thread safe, and then we can open an issue to make a better handling of `dp_stereoAtoms`, as @greglandrum suggested.